### ISSUE TITLE
[Bug]: vllm doesn't support multi-instance GPU

### ISSUE THREAD

**Issue by cfhammill** [vllm-project/vllm#6551]
### Your current environment

```
Collecting environment information...
/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/usage/usage_lib.py:19: RuntimeWarning: Failed to read commit hash:
No module named 'vllm.commit_id'
  from vllm.version import __version__ as VLLM_VERSION
Traceback (most recent call last):
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/cuda/__init__.py", line 306, in _lazy_init
    queued_call()
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/cuda/__init__.py", line 174, in _check_capability
    capability = get_device_capability(d)
                 ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/cuda/__init__.py", line 430, in get_device_capability
    prop = get_device_properties(device)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/cuda/__init__.py", line 448, in get_device_properties
    return _get_device_properties(device)  # type: ignore[name-defined]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: device >= 0 && device < num_gpus INTERNAL ASSERT FAILED at "/build/source/aten/src/ATen/cuda/CUDAContext.cpp":50, please report a bug to PyTorch. device=, num_gpus=

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/mnt/data/home/HammillC/documents/llm-platform/collect_env.py", line 728, in <module>
    main()
  File "/mnt/data/home/HammillC/documents/llm-platform/collect_env.py", line 707, in main
    output = get_pretty_env_info()
             ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/data/home/HammillC/documents/llm-platform/collect_env.py", line 702, in get_pretty_env_info
    return pretty_str(get_env_info())
                      ^^^^^^^^^^^^^^
  File "/mnt/data/home/HammillC/documents/llm-platform/collect_env.py", line 548, in get_env_info
    cuda_module_loading=get_cuda_module_loading_config(),
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/data/home/HammillC/documents/llm-platform/collect_env.py", line 485, in get_cuda_module_loading_config
    torch.cuda.init()
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/cuda/__init__.py", line 260, in init
    _lazy_init()
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/cuda/__init__.py", line 312, in _lazy_init
    raise DeferredCudaCallError(msg) from e
torch.cuda.DeferredCudaCallError: CUDA call failed lazily at initialization with error: device >= 0 && device < num_gpus INTERNAL ASSERT FAILED at "/build/source/aten/src/ATen/cuda/CUDAContext.cpp":50, please report a bug to PyTorch. device=, num_gpus=

CUDA call was originally invoked at:

  File "/mnt/data/home/HammillC/documents/llm-platform/collect_env.py", line 16, in <module>
    import torch
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/__init__.py", line 1480, in <module>
    _C._initExtension(manager_path())
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/cuda/__init__.py", line 238, in <module>
    _lazy_call(_check_capability)
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/torch/cuda/__init__.py", line 235, in _lazy_call
    _queued_calls.append((callable, traceback.format_stack()))

```

### 🐛 Describe the bug

Fortunately, the collect environment output clearly illustrates the problem. Multi-instance GPUs require you to set a single MIG slice or full GPU in order to work correctly. To do this you specify 

```
CUDA_VISIBLE_DEVICES=GPU-<hash>
#or
CUDA_VISIBLE_DEVICES=MIG-<hash>
```

but if you try

```
CUDA_VISIBLE_DEVICES=MIG-<hash> nixGL vllm serve NousResearch/Meta-Llama-3-8B-Instruct --dtype auto --api-key token-abc123
```

you get 

```
Traceback (most recent call last):
  File "/nix/store/91r5lqqzw1bv7xvvyn8kc8b3bmlyyimv-python3.11-vllm-0.5.2/bin/.vllm-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/scripts.py", line 148, in main
    args.dispatch_function(args)
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/scripts.py", line 28, in serve
    run_server(args)
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/entrypoints/openai/api_server.py", line 224, in run_server
    if llm_engine is not None else AsyncLLMEngine.from_engine_args(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 444, in from_engine_args
    engine = cls(
             ^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 373, in __init__
    self.engine = self._init_engine(*args, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 520, in _init_engine
    return engine_class(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/engine/llm_engine.py", line 249, in __init__
    self.model_executor = executor_class(
                          ^^^^^^^^^^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/executor/executor_base.py", line 150, in __init__
    super().__init__(model_config, cache_config, parallel_config,
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/executor/executor_base.py", line 46, in __init__
    self._init_executor()
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/executor/gpu_executor.py", line 24, in _init_executor
    self.driver_worker.init_device()
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/worker/worker.py", line 125, in init_device
    _check_if_gpu_supports_dtype(self.model_config.dtype)
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/worker/worker.py", line 353, in _check_if_gpu_supports_dtype
    compute_capability = current_platform.get_device_capability()
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/platforms/cuda.py", line 48, in get_device_capability
    physical_device_id = device_id_to_physical_device_id(device_id)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/ar0r42s7ygg1h1clhr5i3k50zbqyrf8c-python3-3.11.9-env/lib/python3.11/site-packages/vllm/platforms/cuda.py", line 38, in device_id_to_physical_device_id
    return int(physical_device_id)
           ^^^^^^^^^^^^^^^^^^^^^^^
ValueError: invalid literal for int() with base 10: 'MIG-52b4198c-7365-5c41-8e3b-358b2c032d63'
```

I tried to work around this by setting `NVIDIA_VISIBLE_DEVICES` to the MIG slice with `CUDA_VISIBLE_DEVICES=0` but that didn't work.

**Comment by youkaichao** [comment]
MIG can be complicated.

Can you make docker to launch a container with one MIG GPU? I hope docker can provide an environment that we can treat it as a normal GPU.

**Comment by cfhammill** [comment]
Docker is a work around I could use, yes. It would be better if this could be fixed in the library code though. MIG seems to be well supported in most libraries I use, usually `CUDA_VISIBLE_DEVICES=MIG-<hash>` then using `device=0` for everything works.

There may be particulars that make this harder for vllm. I can take a quick look at the code and see if a simple PR is possible.

**Comment by cfhammill** [comment]
I would guess that the issue is here https://github.com/vllm-project/vllm/blob/main/vllm/platforms/cuda.py#L34

and that adding a conditional or try-catch to handle the case where the device ids aren't a list of ints would work. Only one MIG slice is allowed to be used at a time, so in the case you're setting CUDA_VISIBLE_DEVICES to a MIG slice, device_id can always be set to 0.

**Comment by cfhammill** [comment]
PR'd a simple fix for this, in the process of making sure it works.

**Comment by youkaichao** [comment]
We can only maintain code path for common usecases. Future maintainers, highly likely, don't know anything about MIG. This code can be broken at any time, because we don't have any test against it.

If you can hide the heterogeneity through docker, please do it to save our life. thanks.

**Comment by slin1237** [comment]
why are we closing it
are u suggesting that vLLM will not support MIG now and in the future?
we do have use cases on MIG @youkaichao

**Comment by youkaichao** [comment]
If you can hide the heterogeneity through docker, please do it to save our life. thanks.

**Comment by cfhammill** [comment]
> We can only maintain code path for common usecases. Future maintainers, highly likely, don't know anything about MIG. This code can be broken at any time, because we don't have any test against it.

Confirming that the fix I proposed in https://github.com/vllm-project/vllm/pull/6582 works. It's actually not that hard to test this. `CUDA_VISIBLE_DEVICES` supports device UUIDs, you don't even need MIG to exercise this code path. Just use the UUID for device 0 instead. But understand if MIG is too niche to support. 

That being said assuming CUDA_VISIBLE_DEVICES is either an int or a list of ints is a bug, regardless of whether you want to support MIG or not.

**Comment by youkaichao** [comment]
> assuming CUDA_VISIBLE_DEVICES is either an int or a list of ints is a bug

it is the scope vLLM will maintain. We will not support anything out of the scope. for example, if we want full-fledged MIG support, many code path in vLLM with nvml need to be updated to https://docs.nvidia.com/deploy/nvml-api/group__nvmlMultiInstanceGPU.html . That's too much maintenance burden.

TL;DR: please use docker for MIG.

**Comment by cfhammill** [comment]
Should I open a separate issue for not supporting using device UUIDs? If vLLM doesn't wish to properly support CUDA_VISIBLE_DEVICES adding a legible error for users if they try would be helpful. Even going so far as to suggest the docker work around, or to use my patch.

**Comment by youkaichao** [comment]
sure, you can add it here:

https://github.com/vllm-project/vllm/blob/30efe41532b7cd7202d128ffedfc703c4039767f/vllm/envs.py#L186

this is where all vLLM code reads the `CUDA_VISIBLE_DEVICES` environment variable. error here if you find non-integers.

**Comment by darth-c0d3r** [comment]
Hey, do we have a fix for this? My code was running fine on a previous version of vLLM (0.3.2) but on updating it to 0.6.1, I ran into this issue. Is there a work around for this?

**Comment by youkaichao** [comment]
the solution is to use MIG through docker.

**Comment by d-a-v** [comment]
> the solution is to use MIG through docker.

It does not happen to be a proper solution.
Can't this be reopened and receive a second and closer look ?

**Comment by cfhammill** [comment]
I also think it would be worth revisiting this issue. I've been auditing using `uv` for package management and getting vllm to build from source so I could add my patch was a pain. Not every project/team that uses MIG should/can use docker.

**Comment by dsantiago** [comment]
@youkaichao You say to remove heterogeneity via Docker. Any links towards how to do it? Thanks.

**Comment by larcane97** [comment]
@youkaichao 
It seems that Docker isn’t properly hiding heterogeneous MIG devices.
I’m running vLLM in a k8s env using nvidia gpu, and the pod is running with NVIDIA Docker.
In this setup, the process inside the container seems to access the mig devices directly, without any abstraction or filtering of heterogeneous GPUs.

**Comment by dsantiago** [comment]
> [@youkaichao](https://github.com/youkaichao) It seems that Docker isn’t properly hiding heterogeneous MIG devices. I’m running vLLM in a k8s env using nvidia gpu, and the pod is running with NVIDIA Docker. In this setup, the process inside the container seems to access the mig devices directly, without any abstraction or filtering of heterogeneous GPUs.

Yeah, I am trying to make it work but....

```bash
docker run --rm --gpus '"device=0:0,0:1,0:2,0:3"' \
  -p 8000:8000 \
  -e HF_TOKEN=xxxxxxxxxxxxxxxxxx \
  -e HUGGINGFACE_HUB_CACHE=/data/huggingface_hub \
  -e VLLM_DISABLE_COMPILE_CACHE=0 \
  -e CUDA_VISIBLE_DEVICES=0 \
  -e NVIDIA_MIG_CONFIG="all" \
  --mount type=bind,source=/home/opc/.cache/huggingface/hub/,target=/data/huggingface_hub \
  --ipc=host \
  vllm/vllm-openai \
  --model google/gemma-3-27b-it \
  --max-model-len 3000
```

and output of my sharded GPU (just the first one)

![Image](https://github.com/user-attachments/assets/a40f0e92-a2ab-42fb-9dba-ece553f0448d)

**Comment by benlangfeld** [comment]
@youkaichao I am using the vanilla NVIDIA GPU operator in mixed MIG mode, with 1g.12g slices, and I get the following error:

```
RuntimeError: NVML_SUCCESS == r INTERNAL ASSERT FAILED at "/pytorch/c10/cuda/CUDACachingAllocator.cpp":1015, please report a bug to PyTorch.
```

If using the most-recommended GPU sharding strategy from NVIDIA, the biggest GPU manufacturer, using their operator, on kubernetes (the by far number one container orchestration tool) doesn't count as "common use cases" that vLLM should support, I can't imagine what does.

Please re-open this issue, it is not "completed" as it was marked.

**Comment by benlangfeld** [comment]
> the solution is to use MIG through docker.

Please explain more about how this has been shown to work. We have several instances here of proof that it does not.

**Comment by darth-c0d3r** [comment]
I think I found a better solution to make it work for MIG machines (and other cases where CUDA devices are referred by UUIDs instead of indices).

The problem is in the file `/vllm/platforms/cuda.py` where the function `nvmlDeviceGetHandleByIndex(device_id)` is used even when the device ID is the UUID (and not index).

Here's my fix:

First, make a small change to the function `device_id_to_physical_device_id` and remove the forced type casting to int. The output can now be any ID passed via `CUDA_VISIBLE_DEVICES`. This would get rid of the type error.

After this, instead of always calling `nvmlDeviceGetHandleByIndex(device_id)`, call the appropriate function. I created this wrapper.

```python
def nvmlDeviceGetHandleByIndexOrUUID(physical_device_id):
    """
    physical device id can be either index (int)
    or UUID of a device (starts with GPU- or MIG-)
    """

    if physical_device_id.startswith("GPU-") or physical_device_id.startswith("MIG-"):
        return pynvml.nvmlDeviceGetHandleByUUID(physical_device_id)
    else:
        return pynvml.nvmlDeviceGetHandleByIndex(int(physical_device_id))
```

I replaced all instances of `pynvml.nvmlDeviceGetHandleByIndex` with my custom function `nvmlDeviceGetHandleByIndexOrUUID`.

I'm doing more thorough testing, but this seems to have solved the issue for me. I can share the updated file as well.

**Comment by benlangfeld** [comment]
> I think I found a better solution to make it work for MIG machines (and other cases where CUDA devices are referred by UUIDs instead of indices).
> 
> The problem is in the file `/vllm/platforms/cuda.py` where the function `nvmlDeviceGetHandleByIndex(device_id)` is used even when the device ID is the UUID (and not index).
> 
> Here's my fix:
> 
> First, make a small change to the function `device_id_to_physical_device_id` and remove the forced type casting to int. The output can now be any ID passed via `CUDA_VISIBLE_DEVICES`. This would get rid of the type error.
> 
> After this, instead of always calling `nvmlDeviceGetHandleByIndex(device_id)`, call the appropriate function. I created this wrapper.
> 
> def nvmlDeviceGetHandleByIndexOrUUID(physical_device_id):
>     """
>     physical device id can be either index (int)
>     or UUID of a device (starts with GPU- or MIG-)
>     """
> 
>     if physical_device_id.startswith("GPU-") or physical_device_id.startswith("MIG-"):
>         return pynvml.nvmlDeviceGetHandleByUUID(physical_device_id)
>     else:
>         return pynvml.nvmlDeviceGetHandleByIndex(int(physical_device_id))
> I replaced all instances of `pynvml.nvmlDeviceGetHandleByIndex` with my custom function `nvmlDeviceGetHandleByIndexOrUUID`.
> 
> I'm doing more thorough testing, but this seems to have solved the issue for me. I can share the updated file as well.

Unfortunately it seems the problem is not how this gets fixed, but convincing the maintainer that it is indeed broken.

**Comment by prarabdhshukla** [comment]
> I think I found a better solution to make it work for MIG machines (and other cases where CUDA devices are referred by UUIDs instead of indices).
> 
> The problem is in the file `/vllm/platforms/cuda.py` where the function `nvmlDeviceGetHandleByIndex(device_id)` is used even when the device ID is the UUID (and not index).
> 
> Here's my fix:
> 
> First, make a small change to the function `device_id_to_physical_device_id` and remove the forced type casting to int. The output can now be any ID passed via `CUDA_VISIBLE_DEVICES`. This would get rid of the type error.
> 
> After this, instead of always calling `nvmlDeviceGetHandleByIndex(device_id)`, call the appropriate function. I created this wrapper.
> 
> def nvmlDeviceGetHandleByIndexOrUUID(physical_device_id):
>     """
>     physical device id can be either index (int)
>     or UUID of a device (starts with GPU- or MIG-)
>     """
> 
>     if physical_device_id.startswith("GPU-") or physical_device_id.startswith("MIG-"):
>         return pynvml.nvmlDeviceGetHandleByUUID(physical_device_id)
>     else:
>         return pynvml.nvmlDeviceGetHandleByIndex(int(physical_device_id))
> I replaced all instances of `pynvml.nvmlDeviceGetHandleByIndex` with my custom function `nvmlDeviceGetHandleByIndexOrUUID`.
> 
> I'm doing more thorough testing, but this seems to have solved the issue for me. I can share the updated file as well.

I would like to further elaborate on how to patch vLLM to work with systems that use MIG and not docker. However, I feel this patch is not robust to module updates. I seriously think the module maintainers should consider resolving this issue as a part of their future plans. 

First you must replace all calls to `pynvml.nvmlDeviceGetHandleByIndex` with calls to the custom function defined in https://github.com/vllm-project/vllm/issues/6551#issuecomment-3271526964. After this you are supposed to modify the `device_id_to_physical_device_id` function in `vllm/interface.py`, where you need to remove the type cast to int in the return statement. Even when you do this, you will still run into error because nvml will not be able to calculate the compute capability of your MIG GPU. To circumvent this, you need to patch the `nvmlDeviceGetCudaComputeCapability` function in `third_party/pynvml.py` to fallback to PyTorch's device capability API:

```python
def nvmlDeviceGetCudaComputeCapability(handle):
    import torch
    c_major = c_int()
    c_minor = c_int()
    fn = _nvmlGetFunctionPointer("nvmlDeviceGetCudaComputeCapability")
    ret = fn(handle, byref(c_major), byref(c_minor))
    try:
        _nvmlCheckReturn(ret)
        major, minor = c_major.value, c_minor.value
        if (major,minor) == (0,0):
            raise RuntimeError("NVML returned 0.0 for compute capability")
        return (major, minor)
    except Exception:
        dev_index = torch.cuda.current_device()
        return torch.cuda.get_device_capability(dev_index)
```

This patch seems to work fine for me, but my stand remains clear: vLLM maintainers should look into this issue and take it up now or in the future. Some compute clusters are setup using MIG, and not being able to use vLLM on those is simply too bad. I think it would be worth the time to expand the scope of vLLM to solve this issue.

**Comment by darth-c0d3r** [comment]
Hey! Thanks for adding on to my solution. The custom function `nvmlDeviceGetHandleByIndexOrUUID(physical_device_id):` seems not to work for me because `pynvml.nvmlDeviceGetHandleByUUID` doesn't work for MIG UUID, are you able to get it to run?

**Comment by mychele** [comment]
+1

**Comment by yc-LoAndy** [comment]
+1 very helpful if vLLM supports MIG

**Comment by RodriMora** [comment]
Seems like we are never getting MIG support...

**Comment by marcofaltelli** [comment]
+1

**Comment by haliliceylan** [comment]
+1
